packages/vite/src/node/server/middlewares/environmentTransform.ts

```diff
@@ -0,0 +1,77 @@
+import type { Connect } from 'dep-types/connect'
+import type { ViteDevServer } from '..'
+import {
+  ENVIRONMENT_URL_PUBLIC_PATH,
+  NULL_BYTE_PLACEHOLDER,
+} from '../../../shared/constants'
+
+export function environmentTransformMiddleware(
+  server: ViteDevServer,
+): Connect.NextHandleFunction {
+  return async function viteEnvironmentTransformMiddleware(req, res, next) {
+    if (req.method !== 'GET') {
+      return next()
+    }
+
+    let url: string
+    try {
+      url = decodeURI(req.url!).replace(NULL_BYTE_PLACEHOLDER, '\0')
+    } catch (e) {
+      return next(e)
+    }
+
+    if (!url.startsWith(ENVIRONMENT_URL_PUBLIC_PATH)) {
+      return next()
+    }
+
+    const { pathname, searchParams } = new URL(url, 'http://localhost')
+    const environmentName = pathname.slice(
+      ENVIRONMENT_URL_PUBLIC_PATH.length + 1,
+    )
+    const environment = server.environments[environmentName]
+
+    if (!environmentName || !environment) {
+      res.statusCode = 404
+      res.end()
+      return
+    }
+
+    const moduleUrl = searchParams.get('moduleUrl')
+
+    if (!moduleUrl) {
+      res.statusCode = 404
+      res.end()
+      return
+    }
+
+    // TODO: how to check consistently for all environments(?)
+    // currently ignores if the consumer is a `server`
+    // https://github.com/vitejs/vite/blob/main/packages/vite/src/node/server/transformRequest.ts:271
+    // if (!ensureServingAccess(moduleUrl, server, res, next)) {
+    //   return
+    // }
```

Comment:
I need help with this one 🤔 @patak-dev

Reply:
Related to my comment https://github.com/vitejs/vite/pull/18338#issuecomment-2410418122, exposing `fetchModule` from a bridge server would essentially have the same issue, but what I did is to protect the endpoint by requiring key, which is only available to child process https://github.com/hi-ogawa/vite-environment-examples/pull/132

Reply:
If you have a bridge, it's not a fetchable environment anymore, is it? This seems like completely unrelated approach

Reply:
Yeah, it's probably right to say it's a completely independent concept from "fetchable environment" concept.

Bridge server is only to help a few http rpc between environment implementation and runtime side communication, which can include `fetchModule`. Then, independently from such logic, `devEnv.dispatchFetch(request)` can be implemented as proxing the server spawned inside runtime.

For child process specifically, rpc can be implemented with stdio (or if nodejs child process, onmessage/postMessage works too). And if it's miniflare, then it already supports fairly arbitrary communication between main process and runtime, so that's used.

My impression is that even if we provide this middleware to expose `fetchModule`, I would imagine environment implementation would require other communication primitive for other need, so this middleware alone doesn't seem useful in practice.

Reply:
> I would imagine environment implementation would require other communication primitive for other need

For what needs? How would that achieve those needs in production then if `fetch` is not enough?